pbsmetrics/metrics.go

```diff
@@ -0,0 +1,257 @@
+package pbsmetrics
+
+import (
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/rcrowley/go-metrics"
+	"github.com/vrischmann/go-metrics-influxdb"
+	"strings"
+	"time"
+)
+
+// Labels defines the labels that can be attached to the metrics.
+type Labels struct {
+	Source        DemandSource
+	RType         RequestType
+	Adapter       openrtb_ext.BidderName
+	PubID         string // exchange specific ID, so we cannot compile in values
+	Browser       Browser
+	CookieFlag    CookieFlag
+	RequestStatus RequestStatus
+}
+
+// Label typecasting. Se below the type definitions for possible values
+
+// DemandSource : Demand source enumeration
+type DemandSource string
+
+// RequestType : Request type enumeration
+type RequestType string
+
+// Browser type enumeration
+type Browser string
+
+// CookieFlag : User ID cookie exists flag
+type CookieFlag string
+
+// RequestStatus : The request/adapter return status
+type RequestStatus string
+
+// The demand sources
+const (
+	DemandWeb     DemandSource = "web"
+	DemandApp     DemandSource = "app"
+	DemandUnknown DemandSource = "unknown"
+)
+
+// The request types (endpoints)
+const (
+	ReqTypeLegacy RequestType = "legacy"
+	ReqTypeORTB2  RequestType = "openrtb2"
+	ReqTypeAMP    RequestType = "amp"
+)
+
+// Browser flag; at this point we only care about identifying Safari
+const (
+	BrowserSafari Browser = "safari"
+	BrowserOther  Browser = "other"
+)
+
+// Cookie flag
+const (
+	CookieFlagYes     CookieFlag = "2"
+	CookieFlagNo      CookieFlag = "1"
+	CookieFlagUnknown CookieFlag = "0"
+)
+
+// Request/return status
+const (
+	RequestStatusOK      RequestStatus = "ok"
+	RequestStatusErr     RequestStatus = "err"
+	RequestStatusNoBid   RequestStatus = "nobid"   // Only for adapters
+	RequestStatusTimeout RequestStatus = "timeout" // Only for adapters
```

Comment:
I was a little bit split on this. Seems it would be good to limit the number of label structs I create, it reduces clutter for one. In addition, this allows taking the general request struct and simply set the adapter to create the adapter label struct, instantly inheriting all the settings such as Safari and Cookie without any work.

Reply:
Ok. I'm not super convinced, but... I'm not sure of myself to be too stubborn about it either.

In general each it's better if each "thing" in the code only means one thing.

Reply:
Not super convinced of your side either, but decided to go ahead and flip.